### ISSUE TITLE
Timed abort in pause module

### ISSUE THREAD

**Issue by ansibot** [ansible/ansible#30116]
##### ISSUE TYPE

~Feature  Idea~
Bug Report

##### COMPONENT NAME

pause module

##### ANSIBLE VERSION

2.3

##### SUMMARY

From @bobrik on 2015-06-01T10:43:32Z
It would be nice to about sequence if operator does not respond in timeout. Currently `pause` module can only continue if no action taken. Extra flag like `action_on_timeout` with values `abort` and `continue` would do the trick.

Copied from original issue: ansible/ansible-modules-core#1449

**Comment by ansibot** [comment]
From @ansibot on 2015-06-01T10:43:32Z



@tbielawa ping, this issue is waiting for your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by ansibot** [comment]
From @tbielawa on 2015-06-01T10:43:32Z

That's a pretty neat idea. I'll take a swing at hacking something together that could accomplish that.

**Comment by ansibot** [comment]
From @tbielawa on 2015-06-01T10:43:32Z

@bobrik on second thought, I'm not sure I understand what you're asking for
- http://docs.ansible.com/ansible/pause_module.html

Say I have a pause prompt like this

``` yaml
  - name: Test pausing
    pause:
      prompt: Pausing for a moment (forever!)
```

That produces this play output

```
TASK [Test pausing] ************************************************************
[Test pausing]
Pausing for a moment (forever!):
```

That will pause **indefinitely**. If I press `ctrl+c` and then `a` for **abort** then we see this:

```
Press 'C' to continue the play or 'A' to abort 
// I pressed 'a' here
fatal: [localhost]: FAILED! => {"failed": true, "msg": "user requested abort!"}
```

But if I press `ctrl+c` then `c` the playbook will continue immediately:

```
Pausing for a moment (forever!):
Press 'C' to continue the play or 'A' to abort 
// I pressed 'c' here
ok: [localhost]
```

You said in your original post, "Currently pause module can only continue if _no_ action taken." Did you mean to have that "no" there?

If you want to automatically continue after a certain period of time you can use the `minutes` or `seconds` parameters to trigger that. Here is pausing for **5** seconds:

``` yaml
  - name: Test pausing
    pause:
      prompt: Pausing for a pentasecond
      seconds: 5
```

If I run it and then interrupt it with `ctrl+c` then `c` to continue you get behavior like this:

```
TASK [Test pausing] ************************************************************
Pausing for 5 seconds
(ctrl+C then 'C' = continue early, ctrl+C then 'A' = abort)
Press 'C' to continue the play or 'A' to abort 
// I pressed 'c' here
ok: [localhost]
```

So specifically, are you requesting that if the default prompt mechanism is used, one without `seconds` or `minutes` set, you want that to expire after a set period of time? I think that kind of functionality would require (at least) **two** new parameters in the module
- `timeout_seconds` (**int**) - The number of seconds to wait on a non-timed prompt before the action given in `timeout_action` is taken.
- `timeout_action` (**string** one of `continue` or `abort`) - The action to take on pause tasks after `timeout_seconds` has passed and no `seconds` or `minutes` parameters are set. `continue` will automatically resume the run, `abort` will stop the entire operation. **Mutually exclusive** with the `seconds` and `minutes` parameters.

I'm waiting on requester feedback now to confirm these criteria before I work on any patches for this.

**Comment by ansibot** [comment]
From @tbielawa on 2015-06-01T10:43:32Z

@bcoca Can you update the labels to **Waiting on reporter** please?

**Comment by ansibot** [comment]
From @tbielawa on 2015-06-01T10:43:32Z

Nope, sorry @ansibot, this is still waiting on response from reporter.

**Comment by ansibot** [comment]
@ansibot Greetings! Thanks for taking the time to open this issue. In order for the community to handle your issue effectively, we need a bit more information. 

Here are the items we could not find in your description:
- issue type
- ansible version
- component name

Please set the description of this issue with this template:
https://raw.githubusercontent.com/ansible/ansible/devel/.github/ISSUE_TEMPLATE.md

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @samdoran
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by eric-b-hymowitz** [comment]
(I hope I'm doing this right)

##### ISSUE TYPE

 - Feature Idea

##### COMPONENT NAME

pause

##### ANSIBLE VERSION

```
ansible 2.5.5
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/ehymowitz/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.6/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.6.6 (r266:84292, Aug  9 2016, 06:11:56) [GCC 4.4.7 20120313 (Red Hat 4.4.7-17)]
```

It sounds to me like the request is that the `pause` module honor **both** a user acknowledgement **and** a timeout.  For example:

```
- hosts: localhost
  gather_facts: no
  tasks:
  - name: pause for input
    pause:
      prompt: "Type 'yes' to proceed, or anything else to fail"
    register: pr
  - fail:
      msg: "You did not type 'yes'"
    when: pr.user_input != 'yes'
  - debug:
      msg: "You typed 'yes'!  Thank you very much."
```

This play will print a happy message when you type `yes` to the prompt; otherwise, the play will fail.  But it will wait forever for user input.

It would be nice to have something like this:

```
- hosts: localhost
  gather_facts: no
  tasks:
  - name: pause for input
    pause:
      prompt: "Type 'yes' to proceed, or anything else to fail"
      seconds: 30
    register: pr
  - fail:
      msg: "You did not type 'yes'"
    when: (pr.user_input != 'yes') or (pr.delta == 30)
  - debug:
      msg: "You typed 'yes'!  Thank you very much."
```

Currently, however, once the `seconds` attribute is applied, user input is neither echoed nor captured.

**Comment by CJCShadowsan** [comment]
I'm surprised this has been open so long!

A "continue on enter / timeout after defined seconds" functionality for this prompt is an extremely useful feature - Meaning it could read a default, and if a user wants to override it manually they can, else it continues anyway.

Allows for CI to run through as normal, whilst providing a prompt functionality that could override variables in an interactive mode.

**Comment by bcoca** [comment]
@CJCShadowsan  pause wont 'prompt' when there is no tty present, so it should not be an issue in CI

**Comment by bcoca** [comment]
```
ansible -m pause -a 'prompt=uolosdf seconds=30' localhost
```
no input
```
Pausing for 30 seconds
(ctrl+C then 'C' = continue early, ctrl+C then 'A' = abort)
[pause]
uolosdf:
localhost | SUCCESS => {
    "changed": false,
    "delta": 30,
    "echo": true,
    "rc": 0,
    "start": "2022-01-31 12:35:24.180500",
    "stderr": "",
    "stdout": "Paused for 30.0 seconds",
    "stop": "2022-01-31 12:35:54.180611",
    "user_input": ""
}
```
ansible -m pause -a 'prompt=uolosdf seconds=30' localhost
````
input + ctrl+c + C
```
Pausing for 30 seconds
(ctrl+C then 'C' = continue early, ctrl+C then 'A' = abort)
[pause]
uolosdf:
Press 'C' to continue the play or 'A' to abort
localhost | SUCCESS => {
    "changed": false,
    "delta": 5,
    "echo": true,
    "rc": 0,
    "start": "2022-01-31 12:36:24.926058",
    "stderr": "",
    "stdout": "Paused for 5.57 seconds",
    "stop": "2022-01-31 12:36:30.492309",
    "user_input": ""
}
```
so the main issue is lack of user_input when seconds is added .. almost consider this a bug

**Comment by CJCShadowsan** [comment]
It should both be able to sit at that prompt, _and_ honour the seconds setting for timeout - And it should be "seconds idle".

That way, you could cover both interactive prompting and CI/fully unattended runs by pre-defining variables for your plays as defaults (which is what I am currently having to do, which allows an env var when passed to ansible to define QUIET=true to effectively skip the block for prompt, but this is generally a bit hacky for something ansible should do anyway).

**Comment by jborean93** [comment]
Using `failed_when` on the task will be an easy workaround to this problem. You can define whatever conditional such as `failed_when: '{{ task_res.msg != "user requested abort!" }}'`

**Comment by mkrizek** [comment]
Closing per https://github.com/ansible/ansible/pull/87414#issuecomment-5330140464.
